### DISCUSSION TITLE
Yazi as file picker for Helix with `sway` WM (no mulitplexer needed)

### DISCUSSION THREAD

**Discussion by lukeflo** [sxyazi/yazi#1598] [2. Show and tell]
I just created a workaround to use `yazi` as file picker for Helix with `sway` window manager and `foot` terminal on Void Linux *without* the need of an extra terminal multiplexer (not the most common combination, but maybe some day someone is searching for it... :grin: ). Since Void hasn't as many prebuilt setups, this might be helpful for someone with similar needs:

![recorded](https://github.com/user-attachments/assets/67397935-5e56-4d1d-ae09-3249837b7cab)

For Helix PR [#11164](https://github.com/helix-editor/helix/pull/11164) is very helpful too.

## sway config

Inside my sway config I added following line to make floating terminal windows possible:

```config
for_window [app_id = "floating_foot"] floating enable
```

## script

Furthermore, I need a small wrapper script which opens a new `foot` window with the correct `app_id` to make it float. The script is also necessary since it combines the `yazi` run and the following printing of the file path to `stdout` (`fff` is another possible file manager to use with this script):

```bash
#!/usr/bin/env bash

# Send picker as $1, save selected filepath in cache file
# and print filepath to stdout using cat
# -a `floating_foot` sets app_id for floating term window
# -p option to fff saves selected filepath in tmp-file instead of opening
case "$1" in
fff)
    foot -a 'floating_foot' zsh -c 'fff -p' && cat "/home/lukeflo/.config/cache/fff/opened_file"
    ;;
yazi)
    tmp="$(mktemp -t yazi-picker.XXXXXX)"
    foot -a 'floating_foot' yazi --chooser-file "$tmp"
    cat -- "$tmp"
    rm -f -- "$tmp"
    ;;
*)
    echo "Unknown picker"
    ;;
esac
```

## Helix shortcut

Finally, I just added the script call as shortcut to my `config.toml`:

```toml
[keys.normal.space]
"o" = [ ":open %sh{ sway-picker yazi }" ]
```

Now I can simply press `space` `o` in normal mode. A floating window running `yazi` (or `fff`) pops up and I can select a file which will be opened in my current Helix buffer...

**Comment by AnirudhG07** [comment]
Great Thank you so much! Can you make it into a repository, like [yazi.nvim](https://github.com/mikavilpas/yazi.nvim)?? This will allow people to use configuration more easily. And it would be great if you could add it to [yazi-rs](https://yazi-rs.github.io/docs/resources) and also [awesome-yazi](https://github.com/AnirudhG07/awesome-yazi?tab=readme-ov-file#neovim--vim)

**Comment by lukeflo** [comment]
Thanks. Would like to give it a try, but AFAICS all `yazi` plugins mentioned on your linked sites deal with some Lua code which integrates them with `yazi` itself. My code does not interact with `yazi`'s codebase directly, it just uses `sway` and `helix` config settings to get a workaround. Thus, I'm not sure how to handle this. I could put the script into an own repo (`sway-helix.yazi` or similar) together with some install instructions inside the README if that is enough.

**Comment by AnirudhG07** [comment]
It is okay if it is not in Lua, since neovim plugins are based on Lua only. But helix, sway may not use Lua at all. I have not used helix so I am not aware how configurations work there.
Since helix does not have plugin system(afaik) you can name it however u like(I would recommend not using .yazi since it is not a Yazi plugin technically), check out other editor plugins mentioned in awesome-yazi link for more reference.

It would be great if you could provide a way for configurations for opening in split windows, or diff tabs, etc(if that is so in helix), along with various keybindings from within Yazi, just like yazi.nvim has(just an example).

**Reply by lukeflo** [reply]
OK, thanks for clarifying.

The latter is more tricky because of Helix. As of now, Helix API regarding such things is not very flexible. I can open the file in Helix only using command substitution from its internal command line. That's also only possible in a sophisticated way if Helix is built from source integrating an unmerged PR.

Opening the chosen file inside a new buffer, thus, is only possible using a second Helix command not a custom yazi keypress. That is very likely to change in the future when Helix is evolving.

**Reply by AnirudhG07** [reply]
Ok no problem. Looking forward to seeing your repo!
